File: scripts/download_cases.py
import os
import csv
import urllib.request
import urllib.error
import re
import time
import ssl

# SSL証明書のエラーを回避するためのダミーコンテキスト
ssl_context = ssl._create_unverified_context()

# スクリプトは scripts/ フォルダ内に配置されるため、相対パスで1階層上を指定
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
CSV_PATH = os.path.join(SCRIPT_DIR, '../data/cases.csv')
SAVE_DIR = os.path.join(SCRIPT_DIR, '../cases/raw_sources')
ERROR_LOG_PATH = os.path.join(SAVE_DIR, 'download_error.log')

def clean_html(html_content):
    # 不要なタグとその中身を除去
    html_content = re.sub(r'<script[\s\S]*?</script>', '', html_content, flags=re.IGNORECASE)
    html_content = re.sub(r'<style[\s\S]*?</style>', '', html_content, flags=re.IGNORECASE)
    html_content = re.sub(r'<header[\s\S]*?</header>', '', html_content, flags=re.IGNORECASE)
    html_content = re.sub(r'<footer[\s\S]*?</footer>', '', html_content, flags=re.IGNORECASE)
    html_content = re.sub(r'<nav[\s\S]*?</nav>', '', html_content, flags=re.IGNORECASE)
    
    # すべてのタグを除去
    text = re.sub(r'<[\s\S]*?>', '', html_content)
    
    # 連続する改行や空白を整理
    text = re.sub(r'\n\s*\n', '\n\n', text)
    text = text.strip()
    return text

def download_cases():
    os.makedirs(SAVE_DIR, exist_ok=True)
    
    if not os.path.exists(CSV_PATH):
        print(f"Error: CSV file not found at {CSV_PATH}")
        return
        
    with open(CSV_PATH, mode='r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        rows = list(reader)
        
    print(f"Total cases to check: {len(rows)}")
    
    for idx, row in enumerate(rows, start=1):
        adopter = row['adopter']
        service = row['service']
        url = row['url']
        
        # ファイル名に使えない文字を除去
        safe_adopter = re.sub(r'[\\/*?:"<>|]', '_', adopter)
        safe_service = re.sub(r'[\\/*?:"<>|]', '_', service)
        
        filename = f"{idx:03d}_{safe_adopter}_{safe_service}.txt"
        filepath = os.path.join(SAVE_DIR, filename)
        
        # すでにダウンロード済みのデータがある場合はスキップ
        if os.path.exists(filepath) and os.path.getsize(filepath) > 100:
            print(f"[{idx}/{len(rows)}] Skipped (already exists): {filename}")
            continue
            
        if not url or not url.startswith('http'):
            print(f"[{idx}/{len(rows)}] Skipped (Invalid URL): {url}")
            continue
            
        print(f"[{idx}/{len(rows)}] Downloading: {url} -> {filename}")
        
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
        }
        req = urllib.request.Request(url, headers=headers)
        
        try:
            with urllib.request.urlopen(req, timeout=10, context=ssl_context) as response:
                content_type = response.headers.get('Content-Type', '')
                
                # PDFの場合はバイナリでそのまま保存
                if 'application/pdf' in content_type or url.lower().endswith('.pdf'):
                    pdf_filepath = filepath.replace('.txt', '.pdf')
                    with open(pdf_filepath, 'wb') as out_f:
                        out_f.write(response.read())
                    print(f"  Saved as PDF: {pdf_filepath}")
                    with open(filepath, 'w', encoding='utf-8') as out_f:
                        out_f.write(f"[PDF FILE DOWNLOADED: {pdf_filepath}]\nURL: {url}")
                else:
                    charset = 'utf-8'
                    if 'charset=' in content_type.lower():
                        parts = content_type.split('charset=')
                        if len(parts) > 1:
                            charset = parts[-1].strip().split(';')[0]
                    
                    html_bytes = response.read()
                    try:
                        html_content = html_bytes.decode(charset, errors='replace')
                    except Exception:
                        html_content = html_bytes.decode('utf-8', errors='replace')
                        
                    clean_text = clean_html(html_content)
                    
                    with open(filepath, 'w', encoding='utf-8') as out_f:
                        out_f.write(f"URL: {url}\n\n" + clean_text)
                    print(f"  Successfully saved {len(clean_text)} chars.")
                    
        except Exception as e:
            err_msg = f"Error downloading {url}: {str(e)}"
            print(f"  [ERROR] {err_msg}")
            with open(ERROR_LOG_PATH, 'a', encoding='utf-8') as err_f:
                err_f.write(f"{idx:03d} | {adopter} | {service} | {url} | {str(e)}\n")
                
        # 相手サーバー負荷軽減のための3秒スリープ
        time.sleep(3)

if __name__ == '__main__':
    download_cases()

File: scripts/download_cases.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import https from 'https';
import http from 'http';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const CSV_PATH = path.join(__dirname, '../data/cases.csv');
const SAVE_DIR = path.join(__dirname, '../cases/raw_sources');
const ERROR_LOG_PATH = path.join(SAVE_DIR, 'download_error.log');

// CSVパース関数（簡易的なもの。カンマとダブルクォートを適切に処理する）
function parseCSV(csvText) {
  const lines = csvText.split(/\r?\n/);
  if (lines.length === 0) return [];
  
  // ヘッダーの解析（BOMを除去）
  const firstLine = lines[0].replace(/^\uFEFF/, '');
  const headers = parseCSVLine(firstLine);
  const result = [];
  
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const values = parseCSVLine(line);
    const row = {};
    headers.forEach((header, index) => {
      row[header] = values[index] || '';
    });
    result.push(row);
  }
  return result;
}

function parseCSVLine(line) {
  const result = [];
  let current = '';
  let inQuotes = false;
  
  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      result.push(current.trim());
      current = '';
    } else {
      current += char;
    }
  }
  result.push(current.trim());
  
  // ダブルクォートを除去
  return result.map(val => {
    if (val.startsWith('"') && val.endsWith('"')) {
      return val.slice(1, -1).replace(/""/g, '"');
    }
    return val;
  });
}

function cleanHTML(htmlContent) {
  // 不要なタグとその中身を除去
  let text = htmlContent;
  text = text.replace(/<script[\s\S]*?<\/script>/gi, '');
  text = text.replace(/<style[\s\S]*?<\/style>/gi, '');
  text = text.replace(/<header[\s\S]*?<\/header>/gi, '');
  text = text.replace(/<footer[\s\S]*?<\/footer>/gi, '');
  text = text.replace(/<nav[\s\S]*?<\/nav>/gi, '');
  
  // すべてのタグを除去
  text = text.replace(/<[\s\S]*?>/g, ' ');
  
  // 実体参照のデコード（簡易的なもの）
  text = text.replace(/&nbsp;/g, ' ');
  text = text.replace(/&lt;/g, '<');
  text = text.replace(/&gt;/g, '>');
  text = text.replace(/&amp;/g, '&');
  text = text.replace(/&quot;/g, '"');
  
  // 連続する空白・改行の整理
  text = text.replace(/\r/g, '');
  text = text.replace(/\n\s*\n/g, '\n\n');
  text = text.replace(/[ \t]+/g, ' ');
  text = text.trim();
  return text;
}

function downloadUrl(url) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http;
    const options = {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
      },
      timeout: 15000,
      rejectUnauthorized: false // 証明書エラーを回避
    };
    
    const req = client.get(url, options, (res) => {
      // リダイレクトの処理
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        const redirectUrl = new URL(res.headers.location, url).toString();
        resolve(downloadUrl(redirectUrl));
        return;
      }
      
      if (res.statusCode !== 200) {
        reject(new Error(`HTTP status ${res.statusCode}`));
        return;
      }
      
      const contentType = res.headers['content-type'] || '';
      const isPdf = contentType.includes('application/pdf') || url.toLowerCase().endsWith('.pdf');
      
      const chunks = [];
      res.on('data', (chunk) => chunks.push(chunk));
      res.on('end', () => {
        const buffer = Buffer.concat(chunks);
        resolve({ buffer, isPdf });
      });
    });
    
    req.on('error', (err) => reject(err));
    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Timeout'));
    });
  });
}

async function run() {
  if (!fs.existsSync(SAVE_DIR)) {
    fs.mkdirSync(SAVE_DIR, { recursive: true });
  }
  
  if (!fs.existsSync(CSV_PATH)) {
    console.error(`Error: CSV not found at ${CSV_PATH}`);
    return;
  }
  
  const csvText = fs.readFileSync(CSV_PATH, 'utf-8');
  const rows = parseCSV(csvText);
  console.log(`Total cases to check: ${rows.length}`);
  
  for (let i = 0; i < rows.length; i++) {
    const idx = i + 1;
    const row = rows[i];
    const adopter = row.adopter;
    const service = row.service;
    const url = row.url;
    
    if (!url || !url.startsWith('http')) {
      console.log(`[${idx}/${rows.length}] Skipped (Invalid URL): ${url}`);
      continue;
    }
    
    const safeAdopter = adopter.replace(/[\\/*?:"<>|]/g, '_');
    const safeService = service.replace(/[\\/*?:"<>|]/g, '_');
    const filename = `${String(idx).padStart(3, '0')}_${safeAdopter}_${safeService}.txt`;
    const filepath = path.join(SAVE_DIR, filename);
    
    // すでにファイルが存在する場合はスキップ
    if (fs.existsSync(filepath) && fs.statSync(filepath).size > 100) {
      console.log(`[${idx}/${rows.length}] Skipped (already exists): ${filename}`);
      continue;
    }
    
    console.log(`[${idx}/${rows.length}] Downloading: ${url} -> ${filename}`);
    
    try {
      const { buffer, isPdf } = await downloadUrl(url);
      
      if (isPdf) {
        const pdfFilepath = filepath.replace('.txt', '.pdf');
        fs.writeFileSync(pdfFilepath, buffer);
        fs.writeFileSync(filepath, `[PDF FILE DOWNLOADED: ${pdfFilepath}]\nURL: ${url}`);
        console.log(`  Saved as PDF: ${pdfFilepath}`);
      } else {
        const html = buffer.toString('utf-8');
        const text = cleanHTML(html);
        fs.writeFileSync(filepath, `URL: ${url}\n\n${text}`);
        console.log(`  Successfully saved ${text.length} chars.`);
      }
    } catch (err) {
      console.error(`  [ERROR] ${err.message}`);
      fs.appendFileSync(ERROR_LOG_PATH, `${String(idx).padStart(3, '0')} | ${adopter} | ${service} | ${url} | ${err.message}\n`);
    }
    
    // 3秒スリープ
    await new Promise(resolve => setTimeout(resolve, 3000));
  }
  console.log("Done downloading all cases.");
}

run();
